man/ffpre.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpre 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpre \- This program calculates the relative entropy between an
.IR "l" "\-2" 
Markov model estimation of frequency and the observed frequency
of features.  This is used to calculate an upper limit of 
.I l 
to use for phylogenetic inference.

.SH SYNOPSIS
.BI "ffpre [" "OPTION" "] ... [" "FILE" "] ..."
.SH DESCRIPTION
.PP
Given no options the default behavior of 
.B ffpre 
the progam is to generate the relative entropy value (RE)  (also known as the Kullbach-Leibler divergence)
of feature frequencies using 
.IR "l" "=10" 
and estimates of the frequencies using 
.IR "l" "=9" 
and 
.IR "l" "=8" 
to provide the estimate.  To calculate the 
relative entropies for a range of 
.IR "l" "," 
use the shell script supplied
.B ffpreprof
with the FFP package, which calculates a relative entopry profile.
This utility is used to help determine
the proper feature length range to use for phylogenetic inference.
When the relative entropy approaches zero, this indicates that the
frequency of all features can be determined by using the frequencies
of the 
.IR "l-" "1" 
and 
.IR "l-" "2" 
feature frequencies (i.e. features have the
same frequency as slightly shorter words).  This indicates that each 
.IR "l" "," 
.IR "l-" "1," 
and 
.IR "l-" "2" 
feature occurs in the same context within the
sequence.  Therefore using lengths longer than when the RE is zero
provide no additional information. The default assumes nucleic acid
sequence input.  
.SH OPTIONS
.TP
.BI "\-l " "LEN" ", --length=" "LEN"
.RI "Changes the default length of features to " "LEN" ". The
default length is 10.
Maximum length allowed in 40.
.TP
.B \-d, --disable-ry
Disable RY coding to use ATGC coding.
.TP
.B \-r, --no-reverse
Disable reverse complement matching. Applicable to 
nucleotide sequence only.
.TP
.B \-a, --amino 
Input is amino acid sequence.
.TP
.B \-t, --text
Input is a text file. 
.TP
.B \-v, --version
Print program version.
.TP
.B \-h, --help
Help message.
.PP
.SH EXAMPLES
To calculate the relative entropy between the estimate
and observed feature frequencies:
.PP
.CODE ffpre -l 4 test*.fna
.PP
To calculate the RE for a range of lengths use
.BR "ffpreprof" "."
Note that the RE may oscillate above or below zero - but
given long enough
.I l
it will eventually approach zero.  In mathematical terms
the limit of 
.IR "KL" "(" "l" ")"
as
.I l
approaches infinity is zero.
.PP
.SH FUTURE DIRECTIONS
None.
.PP
.SH AUTHOR
This program was written by Gregory E. Sims
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpvprof(1),
.BR ffpvocab(1),
.BR ffpreprof(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpre.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <getopt.h>
#include <math.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include "hashroll.h"
#include "ffpry.h"
#include "utils.h"
#include "vstring.h"
#include "sighandle.h"
#include "../config.h"

char PROG_NAME[FILENAME_MAX];
#define DEFAULT_WORD_LENGTH 10
#define MAX_LENGTH 30
#define CHAR_BUFFER_SIZE 2000

/**< @todo this needs to be check thoroughly to make sure it produces the correct output
*/
void reNuc(FILE * fp);
void reAA(FILE * fp);
void reText(FILE * fp);
void loopRawRe(HASH * h, HASH * h1, HASH * h2, FILE * fp);

char usage_str[] = "Usage: %s [OPTION] ... [FILE]...\n\
This program prints out the relative entropy between an\n\
l-2 Markov model estimation of frequency and the observed\n\
frequency of features\n\n\
Given no options the default behavior of the program is to\n\
generate the relative entropy value (Kullbach-Leibler Divergence)\n\
of feature frequencies using l=10 and estimates of the frequencies\n\
using l=9 and l=8 to provide the estimate\n\
\t-l LEN, --length=LEN\n\
\t-d, --disable\n\
\t-a, --amino\n\
\t-t, --text\n\
\t-v, --version\n\
\t-v, --help\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";


char *weightVector;
int Length = DEFAULT_WORD_LENGTH;
int maxWordSize = MAX_WORD_SIZE;
int Buffsize = CHAR_BUFFER_SIZE;
unsigned int lcount; /**< Number of features counted of length l */
unsigned int l1count; /**< Number of features counted of length l-1 */
unsigned int l2count; /**< Number of featurs counted of length l-2 */

char aflag = 0;
char dflag = 0;
char tflag = 0;
char rflag = 1;
char fflag = 0;			// not used but needed to link with hashroll.c fix
char mflag = 0;			// not used currently, same as fflag

int main(int argc, char **argv)
{
    char **keys;
    int mode = nucleotide;
    int i;
    int opt;
    FILE *fp;
    double Ef;
    double N;
    double kld;
    HASH h, h1, h2;
    int option_index = 0;
    char *r;
    char *s;
    char *t;

    static struct option long_options[] = {
	{"length", required_argument, 0, 'l'},
	{"disable", no_argument, 0, 'd'},
	{"help", no_argument, 0, 'h'},
	{"amino", no_argument, 0, 'a'},
	{"text", no_argument, 0, 't'},
	{"no-reverse", no_argument, 0, 'r'},
	{"version", no_argument, 0, 'v'},
	{0, 0, 0, 0}
    };

    initSignalHandlers();

    strcpy(PROG_NAME,basename( argv[0] ));

    while ((opt = getopt_long(argc, argv, "l:dhatvr",
			      long_options, &option_index)) != -1)

	switch (opt) {
	case 'l':
	    Length = atoi(optarg);
	    break;
	case 'd':
	    dflag = 1;
	    break;
	case 'r':
	    rflag = !rflag;
	    break;
	case 'a':
	    aflag = 1;
	    mode = amino;
	    break;
	case 't':
	    tflag = 1;
	    mode = text;
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    exit(EXIT_FAILURE);
	    break;

	}

    if ( getenv("MAX_WORD_SIZE") ) {
	maxWordSize=atoi(getenv("MAX_WORD_SIZE"));	    
    }

  if (Length > maxWordSize) 
        fatal_msg("%d: Max Word size is : %d",Length,maxWordSize);


// check -l length to make sure it is valid

    if (Length < 3)
	fatal_msg("Feature Length must be 3 or greater\n");

    init(&h, 0, mode, !dflag, rflag, Length);
    init(&h1, 0, mode, !dflag, rflag, Length - 1);
    init(&h2, 0, mode, !dflag, rflag, Length - 2);
// Must now process file arguments

    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
		fatal_msg("%s: No Such file", *argv);
	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

	loopRawRe(&h, &h1, &h2, fp);

	if (fp != stdin)
	    fclose(fp);

    } while (*argv);


    r = (char *) malloc(sizeof(char) * Length);
    s = (char *) malloc(sizeof(char) * Length);
    t = (char *) malloc(sizeof(char) * Length);

//@todo fix to work with amino acids and text

    lcount = sumValues(&h);
    l1count = sumValues(&h1);
    l2count = sumValues(&h2);

    hashKeys(&h, &keys);

    N = (double) l2count / l1count / l1count;


    kld = 0.0;
    for (i = 0; i < h.keyN; i++) {
	r = strncpy(r, keys[i] + 1, Length - 1);
	r[Length - 1] = '\0';
	s = strncpy(s, keys[i], Length - 1);
	s[Length - 1] = '\0';
	t = strncpy(t, keys[i] + 1, Length - 1);
	t[Length - 2] = '\0';
	Ef = (double) hashValNuc(&h1, r) * hashValNuc(&h1, s) / hashValNuc(&h2,
									   t) *
	    N;
	if (isnormal(Ef)) {
	    kld -= (double) Ef *log2(hashValNuc(&h, keys[i]) / Ef / lcount);
	}
    }
    printf("%lf\n", kld);
    // printf("\n");
    // should free keys too;

    // (a+b)/An   



    freeHash(&h);
    freeHash(&h1);
    freeHash(&h2);
    return EXIT_SUCCESS;
}

// Could change this to an array of hashes
void loopRawRe(HASH * h, HASH * h1, HASH * h2, FILE * fp)
{
    struct stat fattr;
    static size_t optimal_size;
    ssize_t nr;
    static char *buf = NULL;

    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fstat error.\n");

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    buf = (char *) malloc(optimal_size);


    // Make hash aware of its own mode amino, nucleotide or text
    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {
	if (aflag) {
	    pushaa(h,  buf, nr,false);
	    pushaa(h1, buf, nr,false);
	    pushaa(h2, buf, nr,false);
	} else if (tflag) {
	    pushtxt(h,  buf, nr);
	    pushtxt(h1, buf, nr);
	    pushtxt(h2, buf, nr);
	} else {
	    pushatgc(h,  buf, nr,false);
	    pushatgc(h1, buf, nr,false);
	    pushatgc(h2, buf, nr,false);
	}


    }
}
